vmm/src/config.rs

```diff
@@ -3333,11 +3399,164 @@ mod tests {
+
+        let mut still_valid_config = valid_config.clone();
+        still_valid_config.platform = Some(PlatformConfig {
+            num_pci_segments: 16,
+            iommu_segments: Some(vec![1, 2, 3]),
+        });
+        still_valid_config.vsock = Some(VsockConfig {
+            iommu: true,
+            pci_segment: 1,
+            ..Default::default()
+        });
+        assert!(still_valid_config.validate().is_ok());
+
+        let mut invalid_config = valid_config.clone();
+        invalid_config.platform = Some(PlatformConfig {
+            num_pci_segments: 16,
+            iommu_segments: Some(vec![1, 2, 3]),
+        });
+        invalid_config.disks = Some(vec![DiskConfig {
+            iommu: false,
+            pci_segment: 1,
+            ..Default::default()
+        }]);
+        assert!(invalid_config.validate().is_err());
+
+        let mut invalid_config = valid_config.clone();
+        invalid_config.platform = Some(PlatformConfig {
+            num_pci_segments: 16,
+            iommu_segments: Some(vec![1, 2, 3]),
+        });
+        invalid_config.net = Some(vec![NetConfig {
+            iommu: false,
+            pci_segment: 1,
+            ..Default::default()
+        }]);
+        assert!(invalid_config.validate().is_err());
+
+        let mut invalid_config = valid_config.clone();
+        invalid_config.platform = Some(PlatformConfig {
+            num_pci_segments: 16,
+            iommu_segments: Some(vec![1, 2, 3]),
+        });
+        invalid_config.pmem = Some(vec![PmemConfig {
+            iommu: false,
+            pci_segment: 1,
+            ..Default::default()
+        }]);
+        assert!(invalid_config.validate().is_err());
+
+        let mut invalid_config = valid_config.clone();
+        invalid_config.platform = Some(PlatformConfig {
+            num_pci_segments: 16,
+            iommu_segments: Some(vec![1, 2, 3]),
+        });
+        invalid_config.devices = Some(vec![DeviceConfig {
+            iommu: false,
+            pci_segment: 1,
+            ..Default::default()
+        }]);
+        assert!(invalid_config.validate().is_err());
+
+        let mut invalid_config = valid_config.clone();
+        invalid_config.platform = Some(PlatformConfig {
+            num_pci_segments: 16,
+            iommu_segments: Some(vec![1, 2, 3]),
+        });
+        invalid_config.vsock = Some(VsockConfig {
+            iommu: false,
+            pci_segment: 1,
+            ..Default::default()
+        });
+        assert!(invalid_config.validate().is_err());
+
+        let mut invalid_config = valid_config.clone();
+        invalid_config.platform = Some(PlatformConfig {
+            num_pci_segments: 16,
+            iommu_segments: Some(vec![1, 2, 3]),
+        });
+        invalid_config.user_devices = Some(vec![UserDeviceConfig {
+            pci_segment: 1,
+            ..Default::default()
+        }]);
+        assert!(invalid_config.validate().is_err());
```

Comment:
The error returned here turned out to be `Err(UserDevicesRequireSharedMemory)` (not the one we expected, e.g. `Err(IommuNotSupported(1))`), which was reported when I accidentally picked this assertion to try `assert_eq!()`.

Reply:
Also, it may worth adding a `still_valid_config` case for `UserDeviceConfig`, where we need to manually set `VmConfig.memory.shared=true`.

Reply:
Thanks!